encoder.go

```diff
@@ -25,12 +25,33 @@ import (
 	"time"
 )
 
-// Encoder is a format-agnostic interface for all log field encoders. It's not
-// safe for concurrent use.
+// Encoder is a format-agnostic interface for all log entry marshalers. Since
+// log encoders don't need to support the same wide range of use cases as
+// general-purpose marshalers, it's possible to make them much faster and
+// lower-allocation.
+//
+// Encoders need not be fully safe for concurrent use, but it should be safe to
```

Comment:
Discussed offline. Conclusions:

* it's not worth separating this into two separate interfaces.
* instead, make some assumptions about the implementation of this interface that allow for clearer docs.
* in a future PR, explore some options for simplifying this interface, which will also simplify the docs.